File: java8-scala/src/java8/Main01.scala
package java8

/** An example of traits. */
object Main01 {
  def main(args: Array[String]) {
    val f = new FooImpl()
    f.sayHi()
    // f.sayBye()
  }

  trait Foo {
    def sayHi()
  }
  
  class FooImpl extends Foo {
    override def sayHi() {
      println("hi");
    }
  }
}

File: java8/src/java8/Main01.java
package java8;

/** Example of default methods. */
public class Main01 {
  public static void main(String[] args) {
    Foo f = new FooImpl();
    f.sayHi();
    f.sayBye();
  }

  public interface Foo {
    void sayHi();
    
    default void sayBye() {
      System.out.println("bye");
    }
  }

  public static class FooImpl implements Foo {
    @Override
    public void sayHi() {
      System.out.println("hi");
    }
  }
}
